tests/integration/conftest.py

```diff
@@ -267,3 +317,15 @@ async def deploy_charm_if_not_deployed(ops_test: OpsTest, charm, app_name: str,
         # if we're running this locally, we need to wait for "waiting"
         # CI however deploys all in a single model, so traefik is active already.
         await ops_test.model.wait_for_idle([app_name], status="active", timeout=1000)
+
+
+async def safe_relate(ops_test: OpsTest, ep1, ep2):
+    # in pytest-operator CI, we deploy all tests in the same model.
```

Comment:
I think deploying all itests in the same model is not a justification but rather something to avoid.
The introduction of `deploy_traefik_if_not_deployed` and `deploy_charm_if_not_deployed` seems to me like an antipattern that should be avoided.

In the past we were trying out [`reset`](https://pythonlibjuju.readthedocs.io/en/latest/api/juju.model.html?highlight=reset#juju.model.Model.reset) but it wasn't very stable. If we add a `preserve` keyword to `reset` perhaps we could try again.

Reply:
Yeah in hindsight I'd avoid all of that. Models are cheap, I keep forgetting that.
Not in this PR though :D